platform/src/main/java/org/stellar/anchor/platform/callback/RestRateIntegration.java

```diff
@@ -68,29 +83,131 @@ public GetRateResponse getRate(GetRateRequest request) throws AnchorException {
+            ServerErrorException.class);
+      }
+
+      if (fee.getAsset().equals(request.getSellAsset())) {
+        // when fee is in sell_asset,
+        // check that sell_amount is equal to price * buy_amount + (fee ?: 0)
+        BigDecimal expected =
+            new BigDecimal(rate.getPrice())
+                .multiply(new BigDecimal(rate.getBuyAmount()))
+                .add(new BigDecimal(fee.getTotal()))
+                .setScale(sellAsset.getSignificantDecimals(), HALF_UP);
+
+        if (new BigDecimal(rate.getSellAmount()).compareTo(expected) != 0) {
+          logErrorAndThrow(
+              "'sell_amount' is not equal to price * buy_amount + (fee?:0) in the GET /rate response",
+              ServerErrorException.class);
+        }
+      } else {
+        // when fee is in buy_asset,
+        // check that sell_amount is equal to price * (buy_amount + (fee ?: 0))
+        BigDecimal expected =
+            new BigDecimal(rate.getPrice())
+                .multiply(new BigDecimal(rate.getBuyAmount()).add(new BigDecimal(fee.getTotal())))
+                .setScale(sellAsset.getSignificantDecimals(), HALF_UP);
+        if (new BigDecimal(rate.getSellAmount()).compareTo(expected) != 0) {
+          logErrorAndThrow(
+              "'sell_amount' is not equal to price * (buy_amount + (fee ?: 0)) in the GET /rate response",
+              ServerErrorException.class);
+        }
+      }
+
+      if (fee.getDetails() != null) {
+        BigDecimal totalFee =
+            new BigDecimal(0)
+                .setScale(Objects.requireNonNull(feeAsset).getSignificantDecimals(), HALF_UP);
+        for (FeeDescription feeDescription : fee.getDetails()) {
+          if (feeDescription.getName() == null) {
+            logErrorAndThrow(
+                "'fee.details.description[?].name' is missing in the GET /rate response",
+                ServerErrorException.class);
+          }
+          if (!isPositiveNumber(feeDescription.getAmount())) {
+            logErrorAndThrow(
+                "'fee.details[?].description.amount' is missing or not a positive number in the GET /rate response",
+                ServerErrorException.class);
+          }
+          totalFee = totalFee.add(new BigDecimal(feeDescription.getAmount()));
+        }
+
+        // check that sell_amount is equal to price * buy_amount + (fee ?: 0)
+        if (totalFee.compareTo(new BigDecimal(fee.getTotal())) != 0) {
+          logErrorAndThrow(
+              "'sell_amount' is not equal to price * buy_amount + (fee ?: 0) to  in the GET /rate response",
+              ServerErrorException.class);
+        }
+      }
+    } else {
+      // when fee is not present, check that sell_amount is equal to price * buy_amount
+      BigDecimal expected =
+          new BigDecimal(rate.getPrice())
+              .multiply(new BigDecimal(rate.getBuyAmount()))
+              .setScale(sellAsset.getSignificantDecimals(), HALF_UP);
```

Comment:
We use FLOOR for total price, can we also use it here for consistency?
https://github.com/stellar/java-stellar-anchor-sdk/blob/develop/core/src/main/java/org/stellar/anchor/sep38/Sep38Service.java#L531

Reply:
See above.